File: python/ndsel/examples/usage.py
"""Runnable example usage of the ndsel Python API.

    cd python/ndsel && uv run python examples/usage.py
"""

import json

from ndsel import Box, NdselError, Point, Points, Slice, normalize, parse


def show(label: str, value: object) -> None:
    print(f"  {label:18} {json.dumps(value)}")


print("# parse a JSON message, then normalize to the canonical transform")
show("slice [10:20:2]", normalize(parse('{"kind":"slice","start":[10],"stop":[20],"step":[2]}')).to_json())

print("\n# build each kind programmatically (ergonomic dataclass builders)")
show("point (4, 7)", normalize(Point(coords=[4, 7])).to_json())
show("box shape 3x4", normalize(Box(shape=[3, 4])).to_json())
show("slice [0:10:2]", normalize(Slice(start=[0], stop=[10], step=[2])).to_json())
show("points", normalize(Points(coords=[[1, 10], [2, 20]])).to_json())
show("transform 3x4", normalize(parse('{"kind":"transform","input_inclusive_min":[0,0],"input_exclusive_max":[3,4]}')).to_json())

print("\n# error handling: stable, machine-readable reason codes")
for bad in (
    '{"kind":"wat"}',
    '{"kind":"slice","start":[0],"stop":[10],"step":[0]}',
    '{"kind":"slice","start":[0],"stop":[10],"step":[-1]}',
):
    try:
        normalize(parse(bad))
    except NdselError as e:
        print(f"  {bad:52} -> {e.reason.value}")

print("\n# edge cases")
show("big i64 (2**60)", normalize(Point(coords=[2**60])).to_json())
show("unbounded box", normalize(Box(inclusive_min=[0], exclusive_max=["+inf"])).to_json())

File: rust/ndsel/examples/usage.rs
//! Runnable example usage of the ndsel Rust API.
//!
//!     cd rust/ndsel && cargo run --example usage
//!
//! Messages can be built programmatically with the ergonomic constructors
//! (`Point::new`, `BoxSel::new().shape(...)`, `Slice::new(...).step(...)`,
//! `Points::new`) — each is `Into<Message>` — or parsed from JSON with `parse`.
//! Either way, `normalize` reduces a message to a canonical `Transform`, which
//! serializes through serde (`serde_json::to_string`).

use ndsel::{BoxSel, NdselError, Point, Points, Slice, Transform, normalize, parse};

fn show(label: &str, result: Result<Transform, NdselError>) {
    match result {
        Ok(t) => println!("  {label:18} {}", serde_json::to_string(&t).unwrap()),
        Err(e) => println!("  {label:18} ERROR {e}"),
    }
}

fn main() {
    println!("# parse a JSON message, then normalize to the canonical transform");
    show("slice [10:20:2]", parse(r#"{"kind":"slice","start":[10],"stop":[20],"step":[2]}"#).and_then(normalize));

    println!("\n# build each kind programmatically (ergonomic constructors)");
    show("point (4, 7)", normalize(Point::new(vec![4, 7]).into()));
    show("box shape 3x4", normalize(BoxSel::new().shape(vec![3, 4]).into()));
    show("slice [0:10:2]", normalize(Slice::new(vec![0], vec![10]).step(vec![2]).into()));
    show("points", normalize(Points::new(vec![vec![1, 10], vec![2, 20]]).into()));
    show("transform 3x4", parse(r#"{"kind":"transform","input_inclusive_min":[0,0],"input_exclusive_max":[3,4]}"#).and_then(normalize));

    println!("\n# error handling: stable, machine-readable reason codes");
    for bad in [
        r#"{"kind":"wat"}"#,
        r#"{"kind":"slice","start":[0],"stop":[10],"step":[0]}"#,
        r#"{"kind":"slice","start":[0],"stop":[10],"step":[-1]}"#,
    ] {
        match parse(bad).and_then(normalize) {
            Ok(_) => println!("  {bad:52} -> OK (unexpected)"),
            Err(e) => println!("  {bad:52} -> {}", e.reason.code()),
        }
    }

    println!("\n# edge cases");
    show("big i64 (2**60)", normalize(Point::new(vec![1152921504606846976]).into()));
    // `±inf` bounds are not expressible through the finite-i64 setters; use parse.
    show("unbounded box", parse(r#"{"kind":"box","inclusive_min":[0],"exclusive_max":["+inf"]}"#).and_then(normalize));
}
